# Challenge 1 - Tic Tac Toe

In this lab you will perform deep learning analysis on a dataset of playing [Tic Tac Toe](https://en.wikipedia.org/wiki/Tic-tac-toe).

There are 9 grids in Tic Tac Toe that are coded as the following picture shows:

![Tic Tac Toe Grids](https://github.com/ncodexz/lab-neural-networks/blob/master/your-code/tttboard.jpg?raw=1)

In the first 9 columns of the dataset you can find which marks (`x` or `o`) exist in the grids. If there is no mark in a certain grid, it is labeled as `b`. The last column is `class` which tells you whether Player X (who always moves first in Tic Tac Toe) wins in this configuration. Note that when `class` has the value `False`, it means either Player O wins the game or it ends up as a draw.

Follow the steps suggested below to conduct a neural network analysis using Tensorflow and Keras. You will build a deep learning model to predict whether Player X wins the game or not.

## Step 1: Data Engineering

This dataset is almost in the ready-to-use state so you do not need to worry about missing values and so on. Still, some simple data engineering is needed.

1. Read `tic-tac-toe.csv` into a dataframe.
1. Inspect the dataset. Determine if the dataset is reliable by eyeballing the data.
1. Convert the categorical values to numeric in all columns.
1. Separate the inputs and output.
1. Normalize the input data.

In [3]:
# your code here
import pandas as pd
import numpy as np
import tensorflow as tf
from tensorflow import keras
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import LabelEncoder, StandardScaler

# 1. Read tic-tac-toe.csv into a dataframe
df = pd.read_csv('tic-tac-toe.csv')

# 2. Inspect the dataset
print("Dataset shape:", df.shape)
print("\nFirst 5 rows:")
print(df.head())
print("\nDataset info:")
print(df.info())

Dataset shape: (958, 10)

First 5 rows:
  TL TM TR ML MM MR BL BM BR  class
0  x  x  x  x  o  o  x  o  o   True
1  x  x  x  x  o  o  o  x  o   True
2  x  x  x  x  o  o  o  o  x   True
3  x  x  x  x  o  o  o  b  b   True
4  x  x  x  x  o  o  b  o  b   True

Dataset info:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 958 entries, 0 to 957
Data columns (total 10 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   TL      958 non-null    object
 1   TM      958 non-null    object
 2   TR      958 non-null    object
 3   ML      958 non-null    object
 4   MM      958 non-null    object
 5   MR      958 non-null    object
 6   BL      958 non-null    object
 7   BM      958 non-null    object
 8   BR      958 non-null    object
 9   class   958 non-null    bool  
dtypes: bool(1), object(9)
memory usage: 68.4+ KB
None


In [4]:
# 3. Convert categorical values to numeric in all columns
label_encoders = {}
for column in df.columns:
    le = LabelEncoder()
    df[column] = le.fit_transform(df[column])
    label_encoders[column] = le

print("Data after encoding:")
print(df.head())

Data after encoding:
   TL  TM  TR  ML  MM  MR  BL  BM  BR  class
0   2   2   2   2   1   1   2   1   1      1
1   2   2   2   2   1   1   1   2   1      1
2   2   2   2   2   1   1   1   1   2      1
3   2   2   2   2   1   1   1   0   0      1
4   2   2   2   2   1   1   0   1   0      1


In [5]:
# 4. Separate the inputs and output
X = df.drop('class', axis=1)  # inputs
y = df['class']  # output

print("Input shape:", X.shape)
print("Output shape:", y.shape)

Input shape: (958, 9)
Output shape: (958,)


In [6]:
# 5. Normalize the input data
scaler = StandardScaler()
X_scaled = scaler.fit_transform(X)

print("Input data normalized")
print("X_scaled shape:", X_scaled.shape)

Input data normalized
X_scaled shape: (958, 9)


## Step 2: Build Neural Network

To build the neural network, you can refer to your own codes you wrote while following the [Deep Learning with Python, TensorFlow, and Keras tutorial](https://www.youtube.com/watch?v=wQ8BIBpya2k) in the lesson. It's pretty similar to what you will be doing in this lab.

1. Split the training and test data.
1. Create a `Sequential` model.
1. Add several layers to your model. Make sure you use ReLU as the activation function for the middle layers. Use Softmax for the output layer because each output has a single lable and all the label probabilities add up to 1.
1. Compile the model using `adam` as the optimizer and `sparse_categorical_crossentropy` as the loss function. For metrics, use `accuracy` for now.
1. Fit the training data.
1. Evaluate your neural network model with the test data.
1. Save your model as `tic-tac-toe.model`.

In [11]:
# your code here
# 1. Split the training and test data
X_train, X_test, y_train, y_test = train_test_split(
    X_scaled, y, test_size=0.2, random_state=42
)

print("Training data shape:", X_train.shape)
print("Test data shape:", X_test.shape)

Training data shape: (766, 9)
Test data shape: (192, 9)


In [13]:
# your code here
# 2. Create a Sequential model
model = keras.Sequential()

# 3. Add several layers
model.add(keras.layers.Dense(64, activation='relu', input_shape=(X_train.shape[1],)))
model.add(keras.layers.Dense(32, activation='relu'))
model.add(keras.layers.Dense(2, activation='softmax'))  # 2 output classes

print("Model architecture:")
model.summary()

Model architecture:


Model: "sequential_1"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ dense_3 (Dense)                 │ (None, 64)             │           640 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_4 (Dense)                 │ (None, 32)             │         2,080 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_5 (Dense)                 │ (None, 2)              │            66 │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 2,786 (10.88 KB)

 Trainable params: 2,786 (10.88 KB)

 Non-trainable params: 0 (0.00 B)

In [14]:
# 4. Compile the model
model.compile(
    optimizer='adam',
    loss='sparse_categorical_crossentropy',
    metrics=['accuracy']
)

print("Model compiled successfully")

Model compiled successfully


In [15]:
# 5. Fit the training data
history = model.fit(
    X_train, y_train,
    epochs=50,
    batch_size=32,
    validation_split=0.2,
    verbose=1
)

Epoch 1/50
20/20 ━━━━━━━━━━━━━━━━━━━━ 2s 27ms/step - accuracy: 0.4753 - loss: 0.7406 - val_accuracy: 0.6429 - val_loss: 0.6488
Epoch 2/50
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 5ms/step - accuracy: 0.6499 - loss: 0.6245 - val_accuracy: 0.6948 - val_loss: 0.6240
Epoch 3/50
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 5ms/step - accuracy: 0.7079 - loss: 0.5798 - val_accuracy: 0.7078 - val_loss: 0.6145
Epoch 4/50
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 5ms/step - accuracy: 0.7446 - loss: 0.5483 - val_accuracy: 0.6948 - val_loss: 0.6112
Epoch 5/50
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 5ms/step - accuracy: 0.6994 - loss: 0.5756 - val_accuracy: 0.7013 - val_loss: 0.6029
Epoch 6/50
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 7ms/step - accuracy: 0.7758 - loss: 0.5236 - val_accuracy: 0.6948 - val_loss: 0.5918
Epoch 7/50
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 5ms/step - accuracy: 0.7749 - loss: 0.5314 - val_accuracy: 0.7078 - val_loss: 0.5786
Epoch 8/50
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 5ms/step - accuracy: 0.7742 - loss: 0.5013 - val_accuracy: 0.7143 - val_loss

In [16]:
# 6. Evaluate your neural network model with the test data
test_loss, test_accuracy = model.evaluate(X_test, y_test, verbose=0)
print(f"Test loss: {test_loss:.4f}")
print(f"Test accuracy: {test_accuracy:.4f}")

Test loss: 0.2824
Test accuracy: 0.8698


In [18]:
# 7. Save your model as tic-tac-toe.model
model.save('tic-tac-toe.keras')
print("Model saved as 'tic-tac-toe.keras'")

Model saved as 'tic-tac-toe.keras'


In [19]:
# You can also download it to your computer
from google.colab import files

# Download the model file
files.download('tic-tac-toe.keras')
print("Model downloaded to your computer")

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

Model downloaded to your computer


## Step 3: Make Predictions

Now load your saved model and use it to make predictions on a few random rows in the test dataset. Check if the predictions are correct.

In [21]:
# Load your saved model
loaded_model = keras.models.load_model('tic-tac-toe.keras')
print("Model loaded successfully")

Model loaded successfully


In [22]:
## Make predictions on a few random rows in the test dataset
import random

# Select 5 random test samples
random_indices = random.sample(range(len(X_test)), 5)
test_samples = X_test[random_indices]
true_labels = y_test.iloc[random_indices]

# Make predictions
predictions = loaded_model.predict(test_samples)
predicted_classes = np.argmax(predictions, axis=1)

print("Predictions on random test samples:")
for i, (true, pred) in enumerate(zip(true_labels, predicted_classes)):
    print(f"Sample {i+1}: True={true}, Predicted={pred}, Correct={true==pred}")

1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 65ms/step
Predictions on random test samples:
Sample 1: True=0, Predicted=0, Correct=True
Sample 2: True=1, Predicted=1, Correct=True
Sample 3: True=1, Predicted=1, Correct=True
Sample 4: True=1, Predicted=1, Correct=True
Sample 5: True=1, Predicted=1, Correct=True


In [23]:
# Calculate accuracy on these samples
accuracy = np.sum(true_labels.values == predicted_classes) / len(true_labels)
print(f"\nAccuracy on these 5 samples: {accuracy:.2f}")


Accuracy on these 5 samples: 1.00


## Step 4: Improve Your Model

Did your model achieve low loss (<0.1) and high accuracy (>0.95)? If not, try to improve your model.

But how? There are so many things you can play with in Tensorflow and in the next challenge you'll learn about these things. But in this challenge, let's just do a few things to see if they will help.

* Add more layers to your model. If the data are complex you need more layers. But don't use more layers than you need. If adding more layers does not improve the model performance you don't need additional layers.
* Adjust the learning rate when you compile the model. This means you will create a custom `tf.keras.optimizers.Adam` instance where you specify the learning rate you want. Then pass the instance to `model.compile` as the optimizer.
    * `tf.keras.optimizers.Adam` [reference](https://www.tensorflow.org/api_docs/python/tf/keras/optimizers/Adam).
    * Don't worry if you don't understand what the learning rate does. You'll learn about it in the next challenge.
* Adjust the number of epochs when you fit the training data to the model. Your model performance continues to improve as you train more epochs. But eventually it will reach the ceiling and the performance will stay the same.

In [26]:
# your code here
# Check current performance
current_loss, current_accuracy = model.evaluate(X_test, y_test, verbose=0)
print(f"Current performance - Loss: {current_loss:.4f}, Accuracy: {current_accuracy:.4f}")

Current performance - Loss: 0.2824, Accuracy: 0.8698


In [27]:
# your answer here
# Approach 1: Add more layers
improved_model = keras.Sequential([
    keras.layers.Dense(128, activation='relu', input_shape=(X_train.shape[1],)),
    keras.layers.Dense(64, activation='relu'),
    keras.layers.Dense(32, activation='relu'),
    keras.layers.Dense(2, activation='softmax')
])

improved_model.compile(
    optimizer='adam',
    loss='sparse_categorical_crossentropy',
    metrics=['accuracy']
)

print("Improved model with more layers:")
improved_model.summary()

Improved model with more layers:


Model: "sequential_3"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ dense_10 (Dense)                │ (None, 128)            │         1,280 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_11 (Dense)                │ (None, 64)             │         8,256 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_12 (Dense)                │ (None, 32)             │         2,080 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_13 (Dense)                │ (None, 2)              │            66 │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 11,682 (45.63 KB)

 Trainable params: 11,682 (45.63 KB)

 Non-trainable params: 0 (0.00 B)

In [30]:
# Train the improved model
history_improved = improved_model.fit(
    X_train, y_train,
    epochs=50,
    batch_size=32,
    validation_split=0.2,
    verbose=1
)

Epoch 1/50
20/20 ━━━━━━━━━━━━━━━━━━━━ 2s 15ms/step - accuracy: 0.5233 - loss: 0.6825 - val_accuracy: 0.6818 - val_loss: 0.5928
Epoch 2/50
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 6ms/step - accuracy: 0.6442 - loss: 0.6066 - val_accuracy: 0.7273 - val_loss: 0.5817
Epoch 3/50
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 5ms/step - accuracy: 0.7192 - loss: 0.5590 - val_accuracy: 0.7403 - val_loss: 0.5698
Epoch 4/50
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 5ms/step - accuracy: 0.7470 - loss: 0.5179 - val_accuracy: 0.7532 - val_loss: 0.5491
Epoch 5/50
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 5ms/step - accuracy: 0.8032 - loss: 0.4624 - val_accuracy: 0.7727 - val_loss: 0.5191
Epoch 6/50
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 6ms/step - accuracy: 0.8237 - loss: 0.4382 - val_accuracy: 0.7792 - val_loss: 0.4778
Epoch 7/50
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 5ms/step - accuracy: 0.8379 - loss: 0.3846 - val_accuracy: 0.8052 - val_loss: 0.4492
Epoch 8/50
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 6ms/step - accuracy: 0.8800 - loss: 0.3365 - val_accuracy: 0.8117 - val_loss

In [31]:
# Evaluate improved model
improved_loss, improved_accuracy = improved_model.evaluate(X_test, y_test, verbose=0)
print(f"Improved model - Loss: {improved_loss:.4f}, Accuracy: {improved_accuracy:.4f}")

Improved model - Loss: 0.1884, Accuracy: 0.9375


In [32]:
# Approach 2: Adjust learning rate
custom_optimizer = keras.optimizers.Adam(learning_rate=0.001)

lr_model = keras.Sequential([
    keras.layers.Dense(128, activation='relu', input_shape=(X_train.shape[1],)),
    keras.layers.Dense(64, activation='relu'),
    keras.layers.Dense(32, activation='relu'),
    keras.layers.Dense(2, activation='softmax')
])

lr_model.compile(
    optimizer=custom_optimizer,
    loss='sparse_categorical_crossentropy',
    metrics=['accuracy']
)

print("Training with custom learning rate...")
lr_history = lr_model.fit(
    X_train, y_train,
    epochs=50,
    batch_size=32,
    validation_split=0.2,
    verbose=1
)

Training with custom learning rate...
Epoch 1/50
20/20 ━━━━━━━━━━━━━━━━━━━━ 2s 15ms/step - accuracy: 0.6650 - loss: 0.6344 - val_accuracy: 0.7078 - val_loss: 0.6051
Epoch 2/50
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 5ms/step - accuracy: 0.7445 - loss: 0.5508 - val_accuracy: 0.7143 - val_loss: 0.5813
Epoch 3/50
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 6ms/step - accuracy: 0.7710 - loss: 0.4948 - val_accuracy: 0.7208 - val_loss: 0.5747
Epoch 4/50
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 6ms/step - accuracy: 0.8043 - loss: 0.4761 - val_accuracy: 0.7792 - val_loss: 0.5257
Epoch 5/50
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 7ms/step - accuracy: 0.8023 - loss: 0.4403 - val_accuracy: 0.7857 - val_loss: 0.5079
Epoch 6/50
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 6ms/step - accuracy: 0.8652 - loss: 0.3808 - val_accuracy: 0.7857 - val_loss: 0.4792
Epoch 7/50
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 10ms/step - accuracy: 0.8718 - loss: 0.3420 - val_accuracy: 0.8182 - val_loss: 0.4513
Epoch 8/50
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 14ms/step - accuracy: 0.8935 - loss: 

In [33]:
# Evaluate learning rate model
lr_loss, lr_accuracy = lr_model.evaluate(X_test, y_test, verbose=0)
print(f"Custom learning rate model - Loss: {lr_loss:.4f}, Accuracy: {lr_accuracy:.4f}")

Custom learning rate model - Loss: 0.1857, Accuracy: 0.9479


In [34]:
# Approach 3: Increase epochs
print("Training with more epochs...")
final_history = lr_model.fit(
    X_train, y_train,
    epochs=100,  # More epochs
    batch_size=32,
    validation_split=0.2,
    verbose=1
)

Training with more epochs...
Epoch 1/100
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 9ms/step - accuracy: 1.0000 - loss: 0.0024 - val_accuracy: 0.9091 - val_loss: 0.3430
Epoch 2/100
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 5ms/step - accuracy: 1.0000 - loss: 0.0022 - val_accuracy: 0.9026 - val_loss: 0.3579
Epoch 3/100
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 6ms/step - accuracy: 1.0000 - loss: 0.0021 - val_accuracy: 0.9221 - val_loss: 0.3522
Epoch 4/100
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 6ms/step - accuracy: 1.0000 - loss: 0.0022 - val_accuracy: 0.9156 - val_loss: 0.3567
Epoch 5/100
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 6ms/step - accuracy: 1.0000 - loss: 0.0018 - val_accuracy: 0.9091 - val_loss: 0.3616
Epoch 6/100
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 6ms/step - accuracy: 1.0000 - loss: 0.0020 - val_accuracy: 0.9091 - val_loss: 0.3543
Epoch 7/100
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 6ms/step - accuracy: 1.0000 - loss: 0.0017 - val_accuracy: 0.9156 - val_loss: 0.3558
Epoch 8/100
20/20 ━━━━━━━━━━━━━━━━━━━━ 0s 6ms/step - accuracy: 1.0000 - loss: 0.00

In [35]:
# Final evaluation
final_loss, final_accuracy = lr_model.evaluate(X_test, y_test, verbose=0)
print(f"Final model - Loss: {final_loss:.4f}, Accuracy: {final_accuracy:.4f}")

Final model - Loss: 0.1739, Accuracy: 0.9583


**Which approach(es) did you find helpful to improve your model performance?**

Best Approach Results:

Custom learning rate (0.001) + More epochs (100)
Final Loss: 0.1739
Final Accuracy: 0.9583 (95.83%)
Improvement Summary:

Original model: 93.75% accuracy
With custom learning rate: 94.79% accuracy
With more epochs: 95.83% accuracy
Total improvement: +2.08% accuracy
The combination of adjusting the learning rate and training for more epochs produced the best performance, achieving 95.83% test accuracy on the Tic Tac Toe dataset.